src/server/worker/workerrequests.js

```diff
@@ -95,6 +97,55 @@ function WorkerRequests(mainLogger, gmeConfig) {
         return deferred.promise.nodeify(callback);
     }
 
+    function _checkNodeForGuidCollision(core, node, guidsSoFar) {
+        var guid = core.getGuid(node),
+            newGuid;
+
+        if (guidsSoFar[guid]) {
+            newGuid = GUID().replace(/-/g, ''); //TODO should we need global utility for this??
+            logger.info('new guid [' + newGuid + '] has been generated for node [' + core.getPath(node) + ']');
+            core.setAttribute(node, CORE_CONSTANTS.OWN_GUID, newGuid);
+            guidsSoFar[core.getGuid(node)] = true;
+        } else {
+            guidsSoFar[guid] = true;
+        }
+    }
+
+    function _checkGuidsNodeByNode(core, root) {
+        var taskQueue = [''],
+            working = false,
+            timerId,
+            deferred = Q.defer(),
+            guids = {};
+
+        timerId = setInterval(function () {
```

Comment:
I am not sure how to respond to this question:
- I think it is nothing libraryish, but we used it multiple times as this is the only real replacement for recursion (which could be an issue if you have too deep hierarchy)